orchagent/portsorch.cpp

```diff
@@ -456,114 +444,149 @@ bool PortsOrch::initializePort(Port &p)
     return true;
 }
 
-bool PortsOrch::setupVlan(sai_vlan_id_t vlan_id, sai_object_id_t port_id, sai_object_id_t &vlan_member_id)
+bool PortsOrch::addHostIntfs(sai_object_id_t id, string alias, sai_object_id_t &host_intfs_id)
 {
     SWSS_LOG_ENTER();
 
-    sai_status_t status;
+    sai_attribute_t attr;
+    vector<sai_attribute_t> attrs;
+
+    attr.id = SAI_HOSTIF_ATTR_TYPE;
+    attr.value.s32 = SAI_HOSTIF_TYPE_NETDEV;
+    attrs.push_back(attr);
+
+    attr.id = SAI_HOSTIF_ATTR_RIF_OR_PORT_ID;
+    attr.value.oid = id;
+    attrs.push_back(attr);
 
-    status = sai_vlan_api->create_vlan(vlan_id);
+    attr.id = SAI_HOSTIF_ATTR_NAME;
+    strncpy((char *)&attr.value.chardata, alias.c_str(), HOSTIF_NAME_SIZE);
+    attrs.push_back(attr);
+
+    sai_status_t status = sai_hostif_api->create_hostif(&host_intfs_id, attrs.size(), attrs.data());
     if (status != SAI_STATUS_SUCCESS)
     {
-        SWSS_LOG_ERROR("Failed to create VLAN vid:%hu\n", vlan_id);
+        SWSS_LOG_ERROR("Failed to create host interface\n");
         return false;
     }
 
-    sai_attribute_t attr;
-    vector<sai_attribute_t> attrs;
+    return true;
+}
 
-    attr.id = SAI_VLAN_MEMBER_ATTR_VLAN_ID;
-    attr.value.u16 = vlan_id;
-    attrs.push_back(attr);
+bool PortsOrch::addVlan(string vlan_alias)
+{
+    SWSS_LOG_ENTER();
 
-    attr.id = SAI_VLAN_MEMBER_ATTR_PORT_ID;
-    attr.value.oid = port_id;
-    attrs.push_back(attr);
+    sai_vlan_id_t vlan_id = 8888 + stoi(vlan_alias.substr(4));
+    sai_status_t status = sai_vlan_api->create_vlan(vlan_id);
 
-    status = sai_vlan_api->create_vlan_member(&vlan_member_id, attrs.size(), attrs.data());
     if (status != SAI_STATUS_SUCCESS)
     {
-        SWSS_LOG_ERROR("Failed to create VLAN member vid:%hu pid:%llx\n", vlan_id, port_id);
+        SWSS_LOG_ERROR("Failed to create VLAN %s vid:%hu", vlan_alias.c_str(), vlan_id);
+        return false;
+    }
+
+    SWSS_LOG_NOTICE("Create an empty VLAN %s vid:%hu", vlan_alias.c_str(), vlan_id);
+
+    Port vlan(vlan_alias, Port::VLAN);
+    vlan.m_vlan_id = vlan_id;
+    vlan.m_members = set<string>();
+    m_portList[vlan_alias] = vlan;
+
+    return true;
+}
+
+bool PortsOrch::removeVlan(Port vlan)
+{
+    SWSS_LOG_ENTER();
+
+    /* Retry when the VLAN still has members */
+    if (vlan.m_members.size() > 0)
+    {
+        SWSS_LOG_ERROR("Failed to remove non-empty VLAN %s", vlan.m_alias.c_str());
         return false;
     }
```

Comment:
In this case, the OA shall receive first the removing VLAN members notification and then this notification.